services/cd-service/pkg/repository/transformer.go

```diff
@@ -2140,32 +2270,170 @@ func (c *envReleaseTrain) Transform(
+					SkipCause:     nil,
+					Error:         grpc.PublicError(ctx, fmt.Errorf("application %q does not have a version deployed in env %q: %w", appName, upstreamEnvName, err)),
+					AppsPrognoses: nil,
+				}
 			}
 			if upstreamVersion == nil {
-				skipped = append(skipped, fmt.Sprintf(
-					"skipping because there is no version for application %q in env %q \n",
-					appName, upstreamEnvName))
+				appsPrognoses[appName] = ReleaseTrainApplicationPrognosis{
+					SkipCause: &api.ReleaseTrainApplicationPrognosis_SkipCause{
+						SkipCause: api.ReleaseTrainApplicationSkipCause_APP_HAS_NO_VERSION_IN_UPSTREAM_ENV,
+					},
+					Version: 0,
+				}
 				continue
 			}
 			versionToDeploy = *upstreamVersion
 		}
 		if currentlyDeployedVersion != nil && *currentlyDeployedVersion == versionToDeploy {
-			skipped = append(skipped, fmt.Sprintf(
-				"skipping %q because it is already in the version %d\n",
-				appName, *currentlyDeployedVersion))
+			appsPrognoses[appName] = ReleaseTrainApplicationPrognosis{
+				SkipCause: &api.ReleaseTrainApplicationPrognosis_SkipCause{
+					SkipCause: api.ReleaseTrainApplicationSkipCause_APP_ALREADY_IN_UPSTREAM_VERSION,
+				},
+				Version: 0,
+			}
+			continue
+		}
+
+		appLocks, err := state.GetEnvironmentApplicationLocks(c.Env, appName)
+
+		if err != nil {
+			return ReleaseTrainEnvironmentPrognosis{
+				SkipCause:     nil,
+				Error:         err,
+				AppsPrognoses: nil,
+			}
+		}
+
+		if len(appLocks) > 0 {
+			appsPrognoses[appName] = ReleaseTrainApplicationPrognosis{
+				SkipCause: &api.ReleaseTrainApplicationPrognosis_SkipCause{
+					SkipCause: api.ReleaseTrainApplicationSkipCause_APP_IS_LOCKED,
+				},
+				Version: 0,
+			}
+			continue
+		}
+
+		fs := state.Filesystem
+
+		releaseDir := releasesDirectoryWithVersion(fs, appName, versionToDeploy)
+		manifest := fs.Join(releaseDir, "environments", c.Env, "manifests.yaml")
+
+		if _, err := fs.Stat(manifest); err != nil {
+			appsPrognoses[appName] = ReleaseTrainApplicationPrognosis{
+				SkipCause: &api.ReleaseTrainApplicationPrognosis_SkipCause{
+					SkipCause: api.ReleaseTrainApplicationSkipCause_APP_DOES_NOT_EXIST_IN_ENV,
+				},
+				Version: 0,
+			}
 			continue
 		}
 
+		appsPrognoses[appName] = ReleaseTrainApplicationPrognosis{
+			SkipCause: nil,
+			Version:   versionToDeploy,
+		}
+	}
+
+	return ReleaseTrainEnvironmentPrognosis{
+		SkipCause:     nil,
+		Error:         nil,
+		AppsPrognoses: appsPrognoses,
+	}
+}
+
+func (c *envReleaseTrain) Transform(
+	ctx context.Context,
+	state *State,
+	t TransformerContext,
+) (string, error) {
+	renderEnvironmentSkipCause := func(SkipCause *api.ReleaseTrainEnvironmentPrognosis_SkipCause) string {
+		envConfig := c.EnvGroupConfigs[c.Env]
+		upstreamEnvName := envConfig.Upstream.Environment
+		ret := ""
+		switch SkipCause.SkipCause {
+		case api.ReleaseTrainEnvironmentSkipCause_ENV_HAS_NO_UPSTREAM:
+			ret = fmt.Sprintf("Environment '%q' does not have upstream configured - skipping.", c.Env)
```

Comment:
Nitpick: Can we just return directly in all the `case`s?
```suggestion
			return fmt.Sprintf("Environment '%q' does not have upstream configured - skipping.", c.Env)
```